File: 0029.divide_two_integers/solution.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        sign = (dividend < 0) == (divisor < 0)  # 判断正负
        a, b, res = abs(dividend), abs(divisor), 0
        while a >= b:
            x = 0
            while a >= b << (x + 1):  # 每次将b左移一位(乘2)，比较大小, 缩小范围
                x += 1
            res += 1 << x
            a -= b << x
        return min(res if sign else -res, 2**31-1)  # 防一手溢出

File: 0029.divide_two_integers/solution.go
package leetcode0029

import (
	"github.com/Ackerr/Algorithms/utils"
	"math"
)


func divide(dividend int, divisor int) int {
	sign := (dividend < 0) == (divisor < 0)
	a, b, res := utils.Abs(dividend), utils.Abs(divisor), 0
	// if i -> int build error, i -> uint leetcode error
	//for i = 31; i >= 0; i--{
	//	if (a >> i) >= b {
	//		res += 1 << i
	//		a -= b << i
	//	}
	//}
	for a >= b {
		var x uint = 0
		for a >= (b << (x + 1)) {
			x++
		}
		res += 1 << x
		a -= b << x
	}
	if !sign {
		res = -res
	}
	return utils.Min(res, math.MaxInt32)
}
